pkg/ipcache/metadata.go

```diff
@@ -345,18 +343,22 @@ func (ipc *IPCache) UpdatePolicyMaps(ctx context.Context, addedIdentities, delet
 	policyImplementedWG.Wait()
 }
 
-// injectLabels will allocate an identity for the given prefix and the given
-// labels. The caller of this function can expect that an identity is newly
-// allocated with reference count of 1 or an identity is looked up and its
-// reference count is incremented.
+// resolveIdentity will either return a previously-allocated identity for the
+// given prefix or allocate a new one corresponding to the labels associated
+// with the specified prefixInfo.
 //
-// The release of the identity must be managed by the caller, except for the
-// case where a CIDR policy exists first and then the kube-apiserver policy is
-// applied. This is because the CIDR identities before the kube-apiserver
-// policy is applied will need to be converted (released and re-allocated) to
-// account for the new kube-apiserver label that will be attached to them. This
-// is a known issue, see GH-17962 below.
-func (ipc *IPCache) injectLabels(ctx context.Context, prefix netip.Prefix, lbls labels.Labels) (*identity.Identity, bool, error) {
+// This function will take an additional reference on the returned identity.
+// The caller *must* ensure that this reference is eventually released via
+// a call to ipc.IdentityAllocator.Release(). Typically this is tied to whether
+// the caller subsequently injects an entry into the BPF IPCache map:
+//   - If the entry is inserted, we assume that the entry will eventually be
+//     removed, and when it is removed, we will remove that reference from the
+//     identity & release the identity.
+//   - If the entry is not inserted (for instance, because the bpf IPCache map
+//     already has the same IP -> identity entry in the map), immediately release
+//     the reference.
+func (ipc *IPCache) resolveIdentity(ctx context.Context, prefix netip.Prefix, info prefixInfo, restoredIdentity identity.NumericIdentity) (*identity.Identity, bool, error) {
```

Comment:
That's correct, yes. The goal there is that if, for instance, we had allocated identities to associate with certain IPs (eg for FQDN policy), then upon startup, ideally we keep the identities corresponding to those IPs the same, since this simplifies the datapath update process.

Background there is that if we change the identities upon restart, then it is more difficult to coordinate the policymap + ipcache content across all of the endpoints (particularly since ipcache is global). Basically for a hitless upgrade, initially the ipcache map would point all of the prefixes to the old identities(ones allocated from the prior agent run). We'd have to update the policymap content to support both the new and old identities , transition all policymaps to include all of these entries, then transition the ipcache maps so that each IP prefix maps to the new identities, then remove all of the policymap entries corresponding to the old identities. Anything less complicated is liable to end up in a situation where for some period a prefix maps to some identity that is not encoded in a policymap at that time.